internal/controller/apiextensions/claim/api.go

```diff
@@ -34,74 +34,56 @@ import (
 
 // Error strings.
 const (
-	errUpdateClaim           = "cannot update composite resource claim"
-	errUpdateComposite       = "cannot update composite resource"
-	errBindClaimConflict     = "cannot bind claim that references a different composite resource"
-	errBindCompositeConflict = "cannot bind composite resource that references a different claim"
-	errGetSecret             = "cannot get composite resource's connection secret"
-	errSecretConflict        = "cannot establish control of existing connection secret"
-	errCreateOrUpdateSecret  = "cannot create or update connection secret"
+	errUpdateClaim          = "cannot update composite resource claim"
+	errBindClaimConflict    = "cannot bind claim that references a different composite resource"
+	errGetSecret            = "cannot get composite resource's connection secret"
+	errSecretConflict       = "cannot establish control of existing connection secret"
+	errCreateOrUpdateSecret = "cannot create or update connection secret"
 )
 
 // An APIBinder binds claims to composites by updating them in a Kubernetes API
 // server.
 type APIBinder struct {
 	client client.Client
-	typer  runtime.ObjectTyper
 }
 
 // NewAPIBinder returns a new APIBinder.
-func NewAPIBinder(c client.Client, t runtime.ObjectTyper) *APIBinder {
-	return &APIBinder{client: c, typer: t}
+func NewAPIBinder(c client.Client) *APIBinder {
+	return &APIBinder{client: c}
 }
 
 // Bind the supplied claim to the supplied composite.
 func (a *APIBinder) Bind(ctx context.Context, cm resource.CompositeClaim, cp resource.Composite) error {
 	existing := cm.GetResourceReference()
-	proposed := meta.ReferenceTo(cp, resource.MustGetKind(cp, a.typer))
-	if existing != nil && !cmp.Equal(existing, proposed, cmpopts.IgnoreFields(corev1.ObjectReference{}, "UID")) {
+	proposed := meta.ReferenceTo(cp, cp.GetObjectKind().GroupVersionKind())
+	equal := cmp.Equal(existing, proposed, cmpopts.IgnoreFields(corev1.ObjectReference{}, "UID"))
+
+	// We refuse to 're-bind' a claim that is already bound to a different
+	// composite resource.
+	if existing != nil && !equal {
```

Comment:
nitpick: I believe if `existing == nil` we could return early with claim update, rather than checking for equality with a nil object, but I think what you have here may actually be more readable.

Reply:
Ah, good point. I don't feel super strongly about either approach being significantly better so laziness wins out and I'm going to keep it as is. ;)